dashboard/design/survey_summary_design.md

```diff
@@ -0,0 +1,161 @@
+- _Single Responsibility Principle_\
+  Each method, class, module should have only 1 clear purpose. In a stronger statement, each should have only one reason to change.
+- _Dependency Inversion Principle_\
+  High-level modules (complex logic) should not depend on low-level modules (primary operations). Both should depend on abstractions. This allows high-level modules to easily switch to different low-level modules implementation.
+- _Template Method Pattern_\
+  Define the skeleton of an algorithm, deferring some steps to other components. E.g. Query module (step 2) could use different implementations of Parser, Indexer and Filter depends on how data is stored in database. The specific implementations of those child components will be send to Query component as input parameters (dependency injection). Strategy pattern enables switching among implementations of the same component.
+- _Strategy Pattern_\
+  Define an interface common to a family of algorithms, e.g. survey result reducers, and make them interchangeable. This lets the algorithm vary independently from clients that use it.
+
+### Build from the end
+We don't have to start from step 1 (Ingest), we can start from step 3 (Summarize) and create adapters to connect it to the current pipeline.
+
+Specifically for CSF survey work, we will create an input adapter to retrieve survey data from the current data model, and an output adapter to convert summary result to format that the current UI view understands. Looking at it from a different angle, these adapters are just concrete implementations of Retriever and Decorator.
+
+
+# Recommendation
+Option 3. Build a new generic survey pipeline, starting from step 3 (Summarize module) first and create adapters to plug it in the current pipeline. This approach allows us to support 2019 CSF survey sooner by taking advantage of the existing work. Then, gradually build out the rest of the pipeline later.
+
+## Rollout plan
+1. Build 5 components just enough to support 2019 surveys: Retriever (for current db), Transformer, Mapper, Reducer and Decorator (for current UI view). They are components in step 3 and 4 of the above diagram.
+2. Build Fetcher (step 1) to download survey submissions from JotForm to our database.
+3. Build Parser, Indexer and Filter (step 4) to allow quick submission search.
+4. Build another implementation of Retriever (in step 3) which will read from the new db created in step 2 & 3.
+5. Build Presenter (in step 4) that can display result from a single query (e.g. summary results of 1 JotForm survey of a specific workshop)
+6. Build Decorator (in step 4) to provide enough information for the Presenter built in step 5.
+7. Switch 2019 surveys to use the new pipeline. It should then be completely independent from the older pipeline.
```

Comment:
I'm confused by step 1 "...just enough to support 2019 surveys," and step 7 "Switch 2019 surveys to use the new pipeline."  What's the difference?

Reply:
In the 1st step, 2019 surveys will use a hybrid of older pipeline and newer pipeline (with just 5 components at that time). In 7th step, 2019 surveys completely switch to the newer pipeline.